DietPi-Software | Add DietPi Dashboard

nonnorm opened the pull request:
It seems fitting that this gets software ID 200.

<!--
Before submitting a pull request:
- Please ensure the target branch is "dev" (active development): https://github.com/MichaIng/DietPi/tree/dev
- Please ensure changes have been tested and verified functional.
-->
**Status**: Testing

**Reference**: https://github.com/MichaIng/DietPi/issues/448

**Commit list/description**:
<!--
- DietPi-Config | Add "Fan control" option to "Performance Options"
![Screenshot](https://xxx.github.com/images/xxx.png)
- DietPi-Config | Add "Fan control" support for Odroid C2
- DietPi-Config | Syntax fix
-->
- DietPi-Software | Add DietPi-Dashboard

nonnorm committed:
v7.8
+ DietPi-Software | Add DietPi-Dashboard

nonnorm committed:
v7.8
+ DietPi-Software | Add DietPi-Dashboard

nonnorm committed:
v7.8
+ DietPi-Software | Add missing 'then'

nonnorm committed:
v7.8
+ DietPi-Software | Add missing 'then'

nonnorm committed:
v7.8
+ DietPi-Software | CodeFactor Fixes

nonnorm committed:
v7.8
+ DietPi-Software | CodeFactor Fixes

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3499,6 +3507,41 @@ _EOF_
+				G_EXEC mv dietpi_dashboard /usr/local/bin/dietpi-dashboard
+			fi
+
+			G_EXEC chmod +x /usr/local/bin/dietpi-dashboard
```
When not shipping a service, a location in PATH is fine for easy access. But we want to ship a systemd unit, right? Then I suggest to install the binary to `/opt/dietpi-dashboard/dietpi-dashboard`, where we may add other files, like a readme, license, optional assets or such things, to have the option open.

nonnorm committed:
v7.8
+ Multiple | Add service for dashboard

nonnorm committed:
v7.8
+ Multiple | Add service for dashboard

nonnorm committed:
v7.8
+ DietPi-Software | Use /opt/dietpi-dashboard instead of /usr/local/bin

nonnorm committed:
v7.8
+ DietPi-Software | Use /opt/dietpi-dashboard instead of /usr/local/bin

nonnorm committed:
v7.8\n+ DietPi-Software | Add missing G_WHIP_MENU call

nonnorm committed:
v7.8\n+ DietPi-Software | Add missing G_WHIP_MENU call

nonnorm committed:
v7.8
+ DietPi-Software | Add missing "done"

nonnorm committed:
v7.8
+ DietPi-Software | Add missing "done"

nonnorm committed:
v7.8
+ DietPi-Software | Add missing break

nonnorm committed:
v7.8
+ DietPi-Software | Add missing break

nonnorm committed:
v7.8
+ DietPi-Software | Don't check nightly link

nonnorm committed:
v7.8
+ DietPi-Software | Don't check nightly link

nonnorm committed:
v7.8
+ DietPi-Software | Remove unzip step

nonnorm committed:
v7.8
+ DietPi-Software | Remove unzip step

nonnorm committed:
v7.8
+ DietPi-Software | Show that install is interactive

nonnorm committed:
v7.8
+ DietPi-Software | Show that install is interactive

nonnorm committed:
v7.8
+ DietPi-Software | Add in config file

nonnorm committed:
v7.8
+ DietPi-Software | Add in config file

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3551,6 +3551,15 @@ _EOF_
 			G_EXEC chmod +x /opt/dietpi-dashboard/dietpi-dashboard
 
+			cat << _EOF_ > /opt/dietpi-dashboard/config.toml
+port = 8080
```
This port is in use already by Airsonic. To avoid conflicts, we need to use a different one, the next free one is `8088`. Also, we should not overwrite the config file if it exists already, to preserve user adjustments:
```sh
[[ -f '/opt/dietpi-dashboard/config.toml' ]] || cat << _EOF_ > /opt/dietpi-dashboard/config.toml
```

Many thanks for adding this, awesome! 👍

nonnorm replied:
8088 seems good, I'll set it as the default as well when nothing is specifically set.

nonnorm committed:
v7.8

+ DietPi-Software | Change port

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3508,6 +3517,74 @@ _EOF_
+
+			G_EXEC chmod +x /opt/dietpi-dashboard/dietpi-dashboard
+
+			cat << _EOF_ > /opt/dietpi-dashboard/config.toml
```
Shall we keep the convention to have any configs (and data) in `/mnt/dietpi_userdata/<name>`? But for now it is probably overkill with then a single file only in each dir, and it would require an additional option to define the config file path... We can move this any time later when more features which write data, or split config files are implemented.

But we shouldn't overwrite existing config files:
```sh
[[ -f '/opt/dietpi-dashboard/config.toml' ]] || cat << _EOF_ > /opt/dietpi-dashboard/config.toml
```

Not a blocker at all, just an additional little hint for `<<` here documents (same below for the systemd unit): If the ending string `_EOF_` is wrapped into single quotes, the here document is taken literally, hence no variables or command substitutions are expanded. I started to hence use them whenever there is no variable/command substitution intended, similarly how I wrap all literal strings in single quotes instead of double quotes, just to be more explicit:
```sh
cat << '_EOF_' > /opt/dietpi-dashboard/config.toml
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -13658,6 +13735,20 @@ _EOF_
+				G_EXEC systemctl disable --now dietpi-dashboard
+				G_EXEC rm /etc/systemd/system/dietpi-dashboard.service
+
+			fi
```
In case `dietpi-services` was used to change process priorities, or a manual override was created:
```sh
[[ -d '/etc/systemd/system/dietpi-dashboard.service.d' ]] && G_EXEC rm -R /etc/systemd/system/dietpi-dashboard.service.d
```
And I personally prefer the service to be disabled/removed first and the executable afterwards. It doesn't make a difference here, but is the better convention to cover cases where the service contains an `ExecStop`, which would then fail.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1199,6 +1199,15 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
 		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/software/cloud/#k3s'
 
+		#------------------
+		software_id=200
```
Btw, `142` would be free (due to CouchPotato removal with v7.7). But actually I like this using the round number 🙂.

MichaIng committed:
v7.8

+ DietPi-Survey_report | Add DietPi-Dashboard

MichaIng committed:
v7.8

+ README | Add DietPi-Dashboard source link

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1199,6 +1199,15 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
+		aSOFTWARE_DESC[$software_id]='Official DietPi web dashboard'
+		aSOFTWARE_CATX[$software_id]=8
+		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/dietpi_tools/#dietpi-dashboard'
+		aSOFTWARE_INTERACTIVE[$software_id]=1
```
Shall we allow non-interactive installs which then default to `Stable`? Also we could add a `dietpi.txt` setting for this.

nonnorm replied:
Sounds good. Would we do this with the `$G_INTERACTIVE` variable? The option will also probably be removed once the dashboard is out of beta.

MichaIng replied:
Our G_WHIPs are cancelled automatically in this case. We could initiate the variable with the setting in `dietpi.txt`, if it is empty assign `Stable`, then show the whiptail and if it is returns a selection override the variable again. So priority is: Whiptail selection > `dietpi.txt` > `Stable`

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3508,6 +3517,74 @@ _EOF_
+			cat << _EOF_ > /opt/dietpi-dashboard/config.toml
+port = 8088
+
+[tls]
```
Sections are nice to structure a config file, on the other hand it makes it more difficult to script config changes. E.g. if there are several section with `enable`, then `grep`/`sed`/`G_CONFIG_INJECT` cannot be easily used anymore to get/set settings.

Probably there are other arguments (e.g. Rust crate wise) to keep using sections, but otherwise I suggest to skip it for now and use instead:
```
tls = false
cert = "/path/to/cert"
key = "/path/to/key"
```
Hence keep using unique keys and structure the config file, if wanted, with comments.

MichaIng committed:
v7.8

+ CHANGELOG | Thanks to @ravenclaw900, we are priod to announce our own official DietPi web interface for monitoring and managing your DietPi system. It is still in beta phase, as we are testing and implementing more features gradually, but we are happy if you would give it a try, "dietpi-software install 200", and give us feedback: https://github.com/MichaIng/DietPi/issues/448

MichaIng committed:
v7.8

+ CHANGELOG | Little extension

MichaIng committed:
v7.8

+ CHANGELOG | Typo and wording

nonnorm committed:
Merge branch dashboard of https://github.com/MichaIng/DietPi into dashboard

nonnorm committed:
v7.8
+ DietPi-Software | Fixes

nonnorm committed:
v7.8
+ DietPi-Dashboard | Allow unattended install

nonnorm committed:
v7.8
+ DietPi-Software | Switch || with &&

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3508,6 +3516,79 @@ _EOF_
+			)
+
+			local choice_required=
+			while :
```
The loop needs to be removed then, the following is sufficient to leave the variable untouched when cancelled (or non-interactive):
```sh
G_WHIP_MENU "${choice_required}Please choose which version of DietPi Dashboard should be installed." && version=$G_WHIP_RETURNED_VALUE
```
Probably we could add:
```
G_WHIP_BUTTON_CANCEL_TEXT=$version
G_WHIP_DEFAULT_ITEM=$version
```
to pre-select the dietpi.txt value/default and replace the "Cancel" button text with the resulting selection when hitting it 🤔.

nonnorm committed:
v7.8
+ DietPi-Software | Remove loop

nonnorm committed:
v7.8
+ DietPi-Services | Add dashboard and align comments

nonnorm committed:
v7.8
+ DietPi-Software | Start service manually

nonnorm committed:
v7.8
+ DietPi-Dashboard | Set cancel button and default item to dietpi.txt setting

nonnorm committed:
v7.8
+ DietPi-Software | Change config file to remove sections

nonnorm committed:
v7.8
+ DietPi.txt | Typo

MichaIng commented on dietpi/dietpi-services:
```diff
@@ -217,23 +217,24 @@ Available services:
 			# SSH
 			aSERVICE_NAME+=('dropbear')
-			aSERVICE_NAME+=('ssh') 				# OpenSSH Server
+			aSERVICE_NAME+=('ssh') 						# OpenSSH Server
```
When a tab is shown like 8 spaces, all is messed up now 😄. Please revert the changes here as we have spacing aligned for 1 tab = 8 spaces on all scripts.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1199,6 +1199,14 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
+		#------------------
+		software_id=200
+
+		aSOFTWARE_NAME[$software_id]='DietPi Dashboard (beta)'
```
Another question: Shall we call it "DietPi-Dashboard" with a dash before "Dash", to align it with other script names, like "DietPi-CloudShell" and "DietPi-VPN"?

nonnorm replied:
Yeah, that will probably match things up better.

MichaIng committed:
v7.8

+ DietPi-Software | DietPi-Dashboard: Minor coding and format

nonnorm committed:
v7.8
+ DietPi-Software | Rename to DietPi-Dashboard

MichaIng committed:
v7.8

+ DietPi-Services | When printing the status, instead of removing a hardcoded number of 11 characters from the start of the line, remove the first field and delimiter. The number of leading spaces has changed with Bullseye, so that the colon of "Active:" was shown unintentionally.

MichaIng committed:
v7.8

+ DietPi-Software | DietPi-Dashboard: Update architecture name for binary downloads: https://github.com/ravenclaw900/DietPi-Dashboard/issues/12
